### PR TITLE
A re-envisioning of TrueMyth to be more idiomatic and native to C♯.

### PR DESCRIPTION
**Pull request by aggieben**
# Overview

The changes here are sweeping, but with the exception of `.Apply` (`.ap` in [true-myth](https://github.com/true-myth/true-myth)), this represents near-parity with the Javascript/TypeScript edition.  Every element is documented (to varying degrees); most of it is a "translation" from the documentation on truemyth.js.org, with omissions and additions where the language differences demanded.  There are a small handful of Javascript functions without equivalencies in C♯ where they _should_ be implemented; these are annotated with "TODO?".

Major changes of note:
 - Methods overloaded: C♯ supports overloading and is the idiomatic way to give names to variants of the same operation.
 - Only concrete implementations of `Result<T,E>` and `Maybe<T>`, rather than `IResult<T,E>` and  `IMaybe<T>` interfaces with concrete implementations `Ok<T,E>`, `Err<T,E>`, `Just<T>` and `Nothing<T>`, respectively.  The primary reason for this is because C♯ explicitly forbids implicit conversions to or from interfaces.
 - Implicit conversion operators.  These were one of the more important things I wanted to get done.  They allow us to avoid the fussiness and verbosity of the type system imposed when using generics.

# Mapping of functions
Unless otherwise indicated, all C♯ methods below are instance methods called directly on `Result<T,E>` or `Maybe<T>`.  Also, this table will be in the documentation as well as here, as GitHub makes this harder to read than it could be.

Table moved to [the user manual](../blob/next/docs/articles/manual.md).

### CODE REVIEW

**Comment by aggieben** [comment]
@chriskrycho I'd say this is ready to review (when you get a chance - I don't plan to merge until next week sometime), even though there are a few gaps left to fill.

**Comment by chriskrycho** [comment]
In general I'm 100% behind this – and most of the decisions here look great.

The only ones I'd suggest revisiting are mapping the `andThen` functions to `Select` instead of `SelectMany`. C♯ developers are familiar with `SelectMany` via LINQ and the signature are the same, because `SelectMany` *is* `bind`/`flatMap`/`chain`/`andThen`.

Otherwise, I think the rest of the changes make sense.

As an aside: note that JS and TS *can* do overloading, although in a very different way, since JS functions can take and correctly handle dynamic numbers of parameters – and TS can actually correctly capture that (and indeed does so for the sake of automatic currying in True Myth).

**Comment by aggieben** [comment]
Yeah, those functions were a bit of a conundrum.  I get that `SelectMany` is the representation of `bind` et al in LINQ, but it's also a bit of a special case because you're always dealing with sequences of things, and you don't ever see `.SelectMany` used like we do here.  e.g., there's no `Tuple<A,U> Tuple<T,U>.SelectMany<A,U>(Func<T,A> mapFn)`.  Relatedly, `SelectMany` semantically refers to sequences, which we're mostly not dealing with here.  

But having said that, I agree that `T Result<T,U>.Select(T defaultValue)` is not the same thing as `Result<A,U> Result<T,U>.Select(Func<T,A> mapFn)`; I just don't know a better name for it that would still be idiomatic.

I'll dig around a bit more and see if I can come up with some alternatives.

**Comment by chriskrycho** [comment]
I entirely agree with everything you just said… but I also think it's equally applicable to `Select`. My original thought was: go with what C♯ users will find idiomatic, even if LINQ chose poorly when it came to naming these things, given that there was, in fact, prior art. 😆

**Comment by aggieben** [comment]
Well, ok, but at least `.Select` is _semantically_ reasonable, even if it's a bit unconventional with respect to the terminology used in other stacks.  `SelectMany` isn't even that, at least not in this context.

I'm staring to incline toward just using the usual monadic terms: `.Map` and `.Bind`, honestly.  But if we're going to keep `.Select` around as an alias, what would you think about `.SelectMap`?  Got any other ideas?  I just don't know if I can live with `.SelectMany`.  It makes me twitch.

**Comment by chriskrycho** [comment]
I think your best bet is *either* to just bite the bullet (maybe with an alias to make it less painful?) and do `SelectMany`, *or* to just use the normal names. My personal take is that `.Select` *doesn't* make sense – any sense! – for the scenario with `Maybe` or `Result`. Ultimately, though, I'll defer to you, @chrisriesgo, and any other C♯-ers who want to chime in here.

There's precedent for aliasing in True Myth: we supply aliases for compatibility and convenience for people migrating from other libraries, so you can use `.andThen`, `.chain`, or `.flatMap`. (We don't do `.bind` because that [already has a meaning in JavaScript][bind].)

[bind]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind

**Comment by aggieben** [comment]
@chriskrycho @chrisriesgo I think I'm ready for you to have another look.  I squashed all the azure pipelines stuff (at least I think I did), and now both `Result` and `Maybe` have `.Map/.Bind` with `.Select/.SelectMany` aliases.

The documentation is generated, but it's not hosted anywhere, and I think I'd prefer to work on that directly on the master branch because it will be easier to deal with github pages - so if there aren't any objections, I'd like to merge this and cut an rc4 release.

File: README.md
Comment (chriskrycho) [inline README.md]: 😬 This TOC needs to be removed in favor of links to the docs, presumably?

File: docs/articles/manual.md
Comment (chriskrycho) [inline docs/articles/manual.md]: 🤔 I think this should probably be `T Maybe<T>.Unwrap(T defaultValue)`.

File: docs/articles/manual.md
Comment (chriskrycho) [inline docs/articles/manual.md]: 🤔 I'm guessing this should actually be `T Maybe<T>.Unwrap(Func<T> elseFn)`.

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]: 🛠  I think I'd suggest just dropping callbacks entirely here. It's super applicable in the TS world, but I can't imagine anyone writing C♯ this way.

> Another common scenario we find ourselves in is dealing with operations which might fail. In C♯, we usually handle these situations with exceptions. But exceptions are unpredictable: …

File: docs/articles/manual.md
Comment (chriskrycho) [inline docs/articles/manual.md]: 🛠 I'd strongly suggest renaming this back to `UnwrapOr` (and per our discussion in Slack I totally get why it's overloaded). At least to me, the `Or` is very helpful in clarifying that you're getting back an _alternative_ to what's wrapped in the context. (And the same for `Result.Unwrap` below.)

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]:
```suggestion
deal with this scenario. A `Result` is either an `Ok` wrapping around a value (like `Just` does)
```

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]:
```suggestion
or an `Err` wrapping around some type defining what went wrong (not like `Nothing`, which has
```

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]: ❌ This one is no longer correct since you updated the functions so this is `AndThen` (`Chain`/`Bind`).

```suggestion
var htmlForPage = value.AndThen(RenderHtml);
```

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]:
```suggestion
// we can't just invoke `WriteOutput` because it doesn't take a `Result`.
```

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]:
```suggestion
When we have a `Result` instance, we can perform tons of operations on whether it’s `Ok` or
```

File: docs/index.md
Comment (chriskrycho) [inline docs/index.md]:
```suggestion
`Err`, just as we could with a `Just` and `Nothing`, until we need the value. Maybe that’s right
```

**Review by chriskrycho — CHANGES_REQUESTED** [review]
This is fantastic! I had one substantive suggestion and a handful of tweaks to docs.

File: README.md
Comment (aggieben) [reply README.md]: Yeah - it will point to itself once it has a place to point to.  It currently doesn't, and can't until we merge (I plan to try to use GH pages to host the docs)

File: docs/articles/manual.md
Comment (aggieben) [reply docs/articles/manual.md]: Yep.  In the code it's `TValue` to make it easier to keep the types straight, but in the docs I went for brevity, so `T`,`U` there.  I guess one slipped through.

File: docs/index.md
Comment (aggieben) [reply docs/index.md]: I intentionally didn't format these as code throughout the docs because there actually aren't any `Ok`,`Err`,`Just`, or `None` types.

File: docs/index.md
Comment (aggieben) [reply docs/index.md]: Ditto from above.  `Err` isn't a thing.

File: docs/index.md
Comment (aggieben) [reply docs/index.md]: Actually, because we have implicit conversion operators, you probably _could_ just invoke `WriteOutput`.  Either way this needs to not say `setDom`.

Comment (chriskrycho) [reply docs/index.md]: Heh, yes. I didn't even notice that part. 😆

File: docs/index.md
Comment (chriskrycho) [reply docs/index.md]: Ahhhhh, right. That's an interesting difference. 👍

**Review by chriskrycho — APPROVED** [review]
:shipit: !

**Pull request merged**
